File: 0072-edit-distance/0072-edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        
        dp = {}

        def backtrack(i, j):

            #Base Case
            if i == len(word1):
                dp[(i, j)] = len(word2) - j
                return dp[(i, j)]
            
            if j == len(word2):
                dp[(i, j)] = len(word1) - i
                return dp[(i, j)]

            if (i, j) in dp:
                return dp[(i, j)]
            
            if word1[i] == word2[j]:
                #They the same, move.
                dp[(i, j)] = backtrack(i + 1, j + 1)
                return dp[(i, j)]
            else:
                #Insert
                insert = backtrack(i, j + 1)
                #Delete
                delete = backtrack(i + 1, j)
                #Replace
                replace = backtrack(i + 1, j + 1)

                dp[(i, j)] = 1 + min(insert, delete, replace)
                return dp[(i, j)]
                
        return backtrack(0, 0)

File: 0072-edit-distance/0072-edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) 
    {
        
        return dfs(word1, word2, 0, 0);

    }

    int dfs(string& word1, string& word2, int i, int j)
    {

        //Base Case
        if (i >= word1.size())
        {
            return word2.size() - j;
        }

        if (j >= word2.size())
        {
            return word1.size() - i;
        }

        if (dp.count(std::pair(i, j)))
        {
            return dp[std::pair(i, j)];
        }

        int ans;

        //Main part of algorithm...
        if (word1[i] == word2[j])
        {
            return dfs(word1, word2, i + 1, j + 1);
        }
        else
        {
            int insert = 1 + dfs(word1, word2, i, j + 1);
            int del = 1 + dfs(word1, word2, i + 1, j);
            int replace = 1 + dfs(word1, word2, i + 1, j + 1);

            ans = std::min({insert, del, replace});    
            dp[std::pair(i, j)] = ans;
            return ans;
        }

    }

private:

    map<std::pair<int, int>, int> dp;
};